File: iOS/FlowWalletKit/Sources/Cache/Cacheable.swift
/// FlowWalletKit - Caching Protocol
///
/// This protocol defines the caching behavior for Flow wallet components,
/// allowing consistent state persistence across the application.

import Foundation
import Flow

/// Wrapper for cached data with timestamp and expiration
public struct CacheWrapper<T: Codable>: Codable {
    /// The actual cached data
    let data: T
    /// When the data was cached
    let timestamp: Date
    /// Time interval after which the cache expires (nil means never)
    let expiresIn: TimeInterval?
    
    /// Whether the cached data has expired
    var isExpired: Bool {
        guard let expiresIn = expiresIn else {
            return false // Never expires
        }
        return Date().timeIntervalSince(timestamp) > expiresIn
    }
}

/// Protocol defining caching behavior for wallet components
public protocol Cacheable {
    /// The type of data being cached
    associatedtype CachedData: Codable
    
    /// Storage mechanism used for caching
    var storage: StorageProtocol { get }
    
    /// Unique identifier for the cached data
    var cacheId: String { get }
    
    /// The data to be cached
    var cachedData: CachedData? { get }
    
    /// Time interval after which the cache expires (nil means never)
    var cacheExpiration: TimeInterval? { get }
    
    /// Cache the current state
    /// - Parameter expiresIn: Optional custom expiration time
    /// - Throws: Error if caching fails
    func cache(expiresIn: TimeInterval?) throws
    
    /// Load state from cache
    /// - Parameter ignoreExpiration: Whether to ignore expiration time
    /// - Throws: Error if loading fails
    /// - Returns: Cached data if available and not expired
    func loadCache(ignoreExpiration: Bool) throws -> CachedData?
    
    /// Delete cached state
    /// - Throws: Error if deletion fails
    func deleteCache() throws
}

/// Default implementation for common caching operations
public extension Cacheable {
    /// Default cache expiration (never expires)
    var cacheExpiration: TimeInterval? {
        nil
    }
    
    /// Default implementation of cache operation
    func cache(expiresIn: TimeInterval? = nil) throws {
        guard let data = cachedData else { return }
        let wrapper = CacheWrapper(
            data: data,
            timestamp: Date(),
            expiresIn: expiresIn ?? cacheExpiration
        )
        let encoded = try JSONEncoder().encode(wrapper)
        try storage.set(cacheId, value: encoded)
    }
    
    /// Default implementation of cache operation without expiration
    func cache() throws {
        try cache(expiresIn: nil)
    }
    
    /// Default implementation of load operation
    func loadCache(ignoreExpiration: Bool = false) throws -> CachedData? {
        guard let data = try storage.get(cacheId) else {
            return nil
        }

        do {
            let wrapper = try JSONDecoder().decode(CacheWrapper<CachedData>.self, from: data)
            // Check expiration unless ignored
            if !ignoreExpiration && wrapper.isExpired {
                try deleteCache()
                return nil
            }
            return wrapper.data
        } catch {
            throw FWKError.cacheDecodeFailed
        }
    }
    
    /// Default implementation of delete operation
    /// Removes the cached data for this instance
    /// - Throws: Error if deletion fails
    func deleteCache() throws {
        try storage.remove(cacheId)
    }
}

File: Android/wallet/src/main/java/io/outblock/wallet/storage/Cacheable.kt
package io.outblock.wallet.storage

import com.google.gson.Gson
import java.util.Date

/**
 * Interface defining caching behavior for wallet components
 * @param T The type of data being cached (must be serializable)
 */
interface Cacheable<T> {
    /**
     * Storage mechanism used for caching
     */
    val storage: StorageProtocol

    /**
     * Unique identifier for the cached data
     */
    val cacheId: String

    /**
     * The data to be cached
     */
    val cachedData: T?

    /**
     * Time interval after which the cache expires (null means never)
     */
    val cacheExpiration: Long?
        get() = null

    /**
     * Cache the current state
     * @param expiresIn Optional custom expiration time in milliseconds
     */
    fun cache(expiresIn: Long? = null) {
        cachedData?.let { data ->
            val wrapper = CacheWrapper(
                data = data,
                timestamp = Date(),
                expiresIn = expiresIn ?: cacheExpiration
            )
            val json = Gson().toJson(wrapper)
            storage.set(cacheId, json.toByteArray())
        }
    }

    /**
     * Load state from cache
     * @param ignoreExpiration Whether to ignore expiration time
     * @return Cached data if available and not expired
     */
    fun loadCache(ignoreExpiration: Boolean = false): T? {
        val data = storage.get(cacheId) ?: return null
        val json = String(data)
        val wrapper = Gson().fromJson(json, CacheWrapper::class.java)
        
        if (!ignoreExpiration && wrapper.isExpired) {
            deleteCache()
            return null
        }
        
        return wrapper.data as? T
    }

    /**
     * Delete cached state
     */
    fun deleteCache() {
        storage.remove(cacheId)
    }
} 